File: Webscraper/src/main/kotlin/scrapers/TasteOfHome.kt
package scrapers

import Recipe
import org.jsoup.nodes.Document

class TasteOfHome : BaseScraper() {
    /**
     * Return the url given a keyword and page number
     */
    override fun getUrlForPage(page: Number, keyWord: String): String {
        return "https://www.tasteofhome.com/recipes/dishes-beverages/page/$page/?s=$keyWord"
    }

    /**
     * Return recipe urls from a given search page
     */
    override fun getRecipeUrlsFromPage(html: Document): List<String> {
        return html.select(".recipe .entry-title a")
            .map { col -> col.attr("href") }.filter { url -> url.contains("recipes") }

    }

    /**
     * Given the html source code, return a recipe object
     */
    override fun parseRecipeHtml(html: Document, url: String): Recipe {
        val recipe = Recipe()

        recipe.title = getFirstOrEmptyText(html, "h1")
        if (recipe.title == "") {
            println(recipe.title)
            return recipe
        }

        recipe.sourceUrl = url

        // Sometimes videos are used instead of images. In that case no image is returned
        recipe.imgUrl = getFirstOrEmptyAttr(html, "div .featured-container img", "src")

        recipe.steps = arrayListOf(*html.select(".recipe-directions__item span")
            .map { col -> col.ownText() }.toTypedArray())

        recipe.ingredients = arrayListOf(*html.select(".recipe-ingredients li")
            .map { ingredient -> ingredient.ownText() }.toTypedArray())

        recipe.totalTime = getFirstOrEmptyText(html, ".total-time p")

        var rating = 0.0
        val recipeNumberInput = html.select(".rating a")
        var recipeNumberString = ""
        if (recipeNumberInput.isNotEmpty()) {
            recipeNumberString = recipeNumberInput[recipeNumberInput.size - 1].ownText()
        }

        if (recipeNumberString == "Add a Review" || recipeNumberInput.isEmpty()) {
            recipe.rating = "Not reviewed"
            recipe.reviewNumber = 0
        } else {
            html.select(".rating a i").forEach {
                val starStatus = it.classNames()
                if (starStatus.contains("dashicons-star-half")) {
                    rating += 0.5
                } else if (starStatus.contains("dashicons-star-filled")) {
                    rating++
                }
            }
            recipe.rating = "$rating out of 5 stars"
            val numLength = recipeNumberString.indexOf(" ")
            if (numLength in 1..6) {
                recipe.reviewNumber =
                    recipeNumberString.substring(0, numLength).toInt()
            }
        }

        return recipe
    }
}

File: backend/src/webscraper/scrapers/TasteOfHome.ts
import { IRecipe } from "@models/recipe-model";
import BaseScraper from "./BaseScraper";
import * as cheerio from "cheerio";
class TasteOfHome extends BaseScraper {
  
  getRecipeUrlsFromPage(doc: string): string[] {
    const html = cheerio.load(doc);

    return html(".recipe .entry-title a")
      .toArray()
      .map((col) => this.attribute(html(col), "href"))
      .filter((url) => url.includes("recipes"));
  }

  getUrlForPage(page: Number, keyword: string): string {
    return `https://www.tasteofhome.com/recipes/dishes-beverages/page/${page}/?s=${keyword}`;
  }

  parseRecipeHtml(doc: string, url: string): IRecipe {
    const recipe = this.returnEmptyRecipe();
    const html = cheerio.load(doc);
    recipe.title = this.getFirstOrEmptyText(doc, "h1");
    if (recipe.title == "") {
      return recipe;
    }
    recipe.sourceUrl = url;

    // Sometimes videos are used instead of images. In that case no image is returned
    recipe.imageUrl = this.getFirstOrEmptyAttr(
      doc,
      "div .featured-container img",
      "src"
    );

    recipe.steps = html(".recipe-directions__item span")
      .toArray()
      .map((col) => this.textContent(html(col)));

    recipe.ingredients = html(".recipe-ingredients li")
      .toArray()
      .map((ingredient) => this.textContent(html(ingredient)));

    recipe.totalTime = this.getFirstOrEmptyText(doc, ".total-time p");

    const recipeNumberInput = html(".rating a").toArray();
    var recipeNumberString = "";
    if (recipeNumberInput.length) {
      recipeNumberString = this.textContent(
        html(recipeNumberInput[recipeNumberInput.length - 1])
      );
    }

    if (
      recipeNumberString == "Add a Review" ||
      recipeNumberInput.length === 0
    ) {
      recipe.rating = "Not reviewed";
      recipe.reviewNumber = 0;
      return recipe;
    }
    var rating = 0.0;
    html(".rating a i")
      .toArray()
      .forEach((it) => {
        var starStatus = html(it);
        if (starStatus.hasClass("dashicons-star-half")) {
          rating += 0.5;
        } else if (starStatus.hasClass("dashicons-star-filled")) {
          rating++;
        }
      });
    recipe.rating = `${rating} out of 5 stars`;
    const numLength = recipeNumberString.indexOf(" ");
    if (1 <= numLength && numLength <= 6) {
      recipe.reviewNumber = parseInt(
        recipeNumberString.substring(0, numLength)
      );
    }
    return recipe;
  }
}

export default TasteOfHome;
